client/auth.go

```diff
@@ -0,0 +1,386 @@
+	return res, err
+}
+
+// Deletes role.
+func (c *authClient) RoleDelete(request *xlineapi.AuthRoleDeleteRequest) (*xlineapi.AuthRoleDeleteResponse, error) {
+	requestWithToken := xlineapi.RequestWithToken{
+		Token: &c.token,
+		RequestWrapper: &xlineapi.RequestWithToken_AuthRoleDeleteRequest{
+			AuthRoleDeleteRequest: request,
+		},
+	}
+	resp, err := c.handleReq(&requestWithToken, false)
+	if err != nil {
+		return nil, err
+	}
+	res := resp.CommandResp.GetAuthRoleDeleteResponse()
+	res.Header.Revision = resp.SyncResp.Revision
+	return res, err
+}
+
+// Grants role permission
+func (c *authClient) RoleGrantPermission(request *xlineapi.AuthRoleGrantPermissionRequest) (*xlineapi.AuthRoleGrantPermissionResponse, error) {
+	if request.Perm == nil {
+		return nil, errors.New("Permission not given")
+	}
+
+	requestWithToken := xlineapi.RequestWithToken{
+		Token: &c.token,
+		RequestWrapper: &xlineapi.RequestWithToken_AuthRoleGrantPermissionRequest{
+			AuthRoleGrantPermissionRequest: request,
+		},
+	}
+	resp, err := c.handleReq(&requestWithToken, false)
+	if err != nil {
+		return nil, err
+	}
+	res := resp.CommandResp.GetAuthRoleGrantPermissionResponse()
+	res.Header.Revision = resp.SyncResp.Revision
+	return res, err
+}
+
+// Revokes role permission
+func (c *authClient) RoleRevokePermission(request *xlineapi.AuthRoleRevokePermissionRequest) (*xlineapi.AuthRoleRevokePermissionResponse, error) {
+	requestWithToken := xlineapi.RequestWithToken{
+		Token: &c.token,
+		RequestWrapper: &xlineapi.RequestWithToken_AuthRoleRevokePermissionRequest{
+			AuthRoleRevokePermissionRequest: request,
+		},
+	}
+	resp, err := c.handleReq(&requestWithToken, false)
+	if err != nil {
+		return nil, err
+	}
+	res := resp.CommandResp.GetAuthRoleRevokePermissionResponse()
+	res.Header.Revision = resp.SyncResp.Revision
+	return res, err
+}
+
+// Send request using fast path
+func (c authClient) handleReq(req *xlineapi.RequestWithToken, useFastPath bool) (*ProposeResponse, error) {
+	proposeId := c.generateProposeId()
+	cmd := xlineapi.Command{Request: req, ProposeId: proposeId}
+
+	if useFastPath {
+		res, err := c.curpClient.propose(&cmd, true)
+		return res, err
+	} else {
+		res, err := c.curpClient.propose(&cmd, false)
+		if err != nil {
+			return res, err
+		} else {
+			if res != nil && res.SyncResp == nil {
+				panic("syncResp is always Some when useFastPath is false")
+			}
+			return res, err
+		}
+	}
+}
+
+// Generate hash of the password
+func (c authClient) hashPassword(password []byte) string {
+	// Generate random salt
+	salt := make([]byte, 16)
+	_, err := rand.Read(salt)
+	// This doesn't seems to be fallible
+	if err != nil {
+		panic(err)
+	}
+	// Generate a hash of the password using PBKDF2
+	hashed := pbkdf2.Key(password, salt, 4096, 32, sha256.New)
+	// Combining salt and hashed into the final hash result
+	hash := append(salt, hashed...)
+	// Formatted as a hex string and returned
+	return fmt.Sprintf("%x", hash)
```

Comment:
It seems that this does not match to the default format of PHC in the xline server.

Have you test it?

FYI: 
- https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md#b64
- https://docs.rs/password-hash/0.5.0/src/password_hash/lib.rs.html#127-199

Reply:
Yes it doesn't match, I've modified and tested it and it matches for now!